### PR TITLE
Document Rails application conventions

### PR DESCRIPTION
**Pull request by kevindew**
tl;dr enjoy [the rendered view](https://github.com/alphagov/govuk-developer-docs/blob/rails-app-conventions/source/manual/conventions-for-rails-applications.html.md) and shout if this is either unhelpful, superfluous or divisive.

This aims to be living documentation that captures the current patterns and approaches GOV.UK uses in Rails applications. I'm expecting it'll only change somewhat gradually, it feels like most of the things written in here could have been catalogued a couple of years ago.

The motivation for putting this together was that it felt like some parts of this were widely known among GOV.UK developers, others less so, but hardly any parts of it are documented anywhere as a preference. This likely leads to teams having to re-evaluate the same decisions' previous teams made. The Publishing Workflow team went through a lot of this, which motivated me to document some of those outcomes.

I've not been too specific on all the why's involved in these, mostly as this would be quite verbose and often the decisions on the preferences are just based on "this worked well for us, why do it differently?"

The things that I think could be a bit divisive here are:

- Using ActiveJob as a preference to using Sidekiq directly. I went for that choice as we've got a number of apps with ActiveJob now, which make use of it for it's ActionMailer integration, which isn't as straight forward with Sidekiq alone. I thought it was better to have a preference of one direction rather than an or. Myself I really don't mind which one is our preference, just that we have a preference.
- Organising code section. I don't think we've had the great app vs lib debate outside the Publishing Workflow team. Though since we did and came to a good resolution it seems worth putting forward as a practice, particularly as it's common across many GOV.UK apps.
- Microcopy editing via content designers. This practice seems to be growing but I don't know if we've ever demonstrated much value from it and still have long running compromises.

It seems that https://docs.publishing.service.gov.uk/manual/which-gem-to-use.html was set-up with a similar purpose but was never flushed out much. I've removed this page as this new one supersedes it.

I did debate about adding some text on patterns we recommend and those we discourage (👍 for service classes, low complexity controllers, expiring rake tasks; 👎 for single-table inheritance, lib/services.rb, presenter classes) but felt those opinions could be somewhat divisive. Thought I'd see how this works out first.

### CODE REVIEW

File: source/manual/conventions-in-rails-applications.html.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+[issue-example]: https://github.com/alphagov/govuk_app_config/issues/121
+
+## Data storage
+
+For non-specialist database needs you should use
+[PostgreSQL](https://www.postgresql.org/) with [ActiveRecord][]. The
+[db/seeds.rb][] can be used to populate the database for development and test
+environments.
+
+For [key-value datastore][] access (such as queues or distributed caches)
```
Comment (theseanything) [inline source/manual/conventions-in-rails-applications.html.md:100]: When should we be using RabbitMQ as a queue?

File: source/manual/conventions-in-rails-applications.html.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+[GOV.UK Notify]: https://www.notifications.service.gov.uk/
+[mail-notify]: https://github.com/dxw/mail-notify
+
+## Configure Rails as `api_only` for API projects
+
+For applications that do not serve requests to a web browser you should
+configure the Rails application to be [API only][].
+
+[API only]: https://guides.rubyonrails.org/api_app.html
```
Comment (theseanything) [inline source/manual/conventions-in-rails-applications.html.md:160]: Maybe not for this PR - but would be useful to have guidance around serving both an API and HTML from the same application - (is that even a thing we should be doing?)

Comment (theseanything) [reply source/manual/conventions-in-rails-applications.html.md:160]: what is our preferred why to serialise JSON - jbuilder, AMS etc...

File: source/manual/conventions-in-rails-applications.html.md (hunk 1 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:100]: Thanks for taking a look @theseanything 😄 

RabbitMQ is used for inter-app communication whereas Redis is only used on a single app basis. I'm not too sure RabbitMQ usage is a convention since I think we only have the one Publishing API queue that is used for communicating between apps (unless I'm missing something). Though I can see how clarifying those different types of queue usage would be useful.

File: source/manual/conventions-in-rails-applications.html.md (hunk 2 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:160]: > Maybe not for this PR - but would be useful to have guidance around serving both an API and HTML from the same application - (is that even a thing we should be doing?)

I guess if you've got a HTML rendering Rails app it's ok to have a few API endpoints, nearly all apps implement this through /healthcheck endpoints. Though too many might be a sign you've got a monolith growing (of course, Whitehall has quite a rich API)

Where I think this smells though is when an application that is essentially API only has HTML endpoints. We have this problem with Publishing API, where we have one, seldomly used, [debug endpoint](https://github.com/alphagov/publishing-api/blob/master/app/controllers/debug_controller.rb) that is the lone requirement for all the extra Rails baggage needed for view rendering etc.





> and what is our preferred way to serialise JSON - jbuilder, AMS etc...

I don't think we've managed to standardise on that, Whitehall uses jbuilder but I don't think anything else does. Instead it seems to get wrapped up in the ambiguous concept of Presenters where in some apps they're used for serializing data and others as a presentation model. It'd be great if our next API driven app could establish precedents for serialisation to move us away from mixing them in with the presenter concept.

Comment (theseanything) [reply source/manual/conventions-in-rails-applications.html.md:160]: Yup agree - I think overall we haven't got any formalisation on guidance around building APIs in Rails. Sorry for opening that can of worms - probably a conversation to have in another place, as it doesn't seem helpful for this PR. I can't think of anything else to add to this section for now - so happy to resolve this convo!

File: source/manual/conventions-in-rails-applications.html.md (hunk 1 above)
Comment (theseanything) [reply source/manual/conventions-in-rails-applications.html.md:100]: Ah yup -  inter-app communication is a good property distinguish it with!

File: source/manual/conventions-in-rails-applications.html.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+[GOV.UK admin layout]: https://components.publishing.service.gov.uk/component-guide/layout_for_admin
+
+## Testing strategies
+
+We aim for well tested Rails applications that provide a good degree of coverage
+and provide assurances that an application is operating correctly. We value
+test suites that are fast to run, easy to read and [test things consistently at
+different abstractions (unit, integration and functional)][cp-testing-strategy].
+
+### Common tooling
```
Comment (ChrisBAshton) [inline source/manual/conventions-in-rails-applications.html.md:231]:
```suggestion
### Common tooling
```

File: source/manual/conventions-in-rails-applications.html.md (hunk 1 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:100]: I'll add some text in before this is merged.

File: source/manual/conventions-in-rails-applications.html.md (hunk 2 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:160]: Gotcha, yes I don't think we've got any clear conventions in the direction of adding an API to an app. We have conventions the other way: for example, Email Alert API is a machine only API and we have a different app, Email Alert Frontend, to provide the users interface. 

I guess though when you hit the situation like you two did with signon, where the system is architectured as a public facing app, then you're pretty snookered to add an API on later without doing it directly to the app. So I wonder if this is more broader architectural convention of creating systems API first. Definitely think this ventures outside of this scope, but good discussion none the less 👍

File: source/manual/conventions-in-rails-applications.html.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+## JavaScript package management
+
+As per Rails conventions GOV.UK applications should use
+[Yarn](https://yarnpkg.com) for JavaScript packages. This is because it
+integrates better with Rails than [NPM](https://www.npmjs.com/).
+
+## Frontend assets
+
+GOV.UK applications use the Rails [asset pipeline][] for building assets. We
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:137]: Yeah, I think that's pretty much it. It's conventional to use webpacker in Rails and we don't and currently can't.

File: source/manual/conventions-in-rails-applications.html.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+[asset pipeline]: https://guides.rubyonrails.org/asset_pipeline.html
+[webpacker]: https://github.com/rails/webpacker
+[blocking issues]: https://github.com/alphagov/govuk_publishing_components/issues/505
+
+## Sending emails
+
+[GOV.UK Notify][] is the preferred approach for sending email. The
+[mail-notify][] gem is conventionally used to integrate Notify with
+[ActionMailer][]. There is a shared GOV.UK Publishing Notify account that can
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:149]: Oh yes, cool, I didn't know about that doc.

File: source/manual/conventions-in-rails-applications.html.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+[Sidekiq]: https://sidekiq.org
+[ActiveJob]: https://guides.rubyonrails.org/active_job_basics.html
+[ActionMailer]: https://guides.rubyonrails.org/action_mailer_basics.html
+[sidekiq-scheduler]: https://github.com/moove-it/sidekiq-scheduler
+
+## JavaScript package management
+
+As per Rails conventions GOV.UK applications should use
+[Yarn](https://yarnpkg.com) for JavaScript packages. This is because it
+integrates better with Rails than [NPM](https://www.npmjs.com/).
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:133]: It is integrated in Rails core so we get rake tasks like yarn:install by default and when you run assets:precompile that'll also run yarn install automatically. Whereas if you wanted to go down an NPM route I believe you have to disable stuff and add your own tasks to achieve the same. 

I'll try find a link on rails guides about the integration so I can show evidence succinctly.

File: source/manual/conventions-in-rails-applications.html.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+## Organising your code
+
+Organising code in a Rails application is frequently a source of debate and
+confusion, [particularly as to what code belongs in the `/app` directory and
+what belongs in the `/lib` directory][app-lib]. We aim to follow what appears
+to be the prevalent convention where:
+
+- `/app` is used for:
+  - the default Rails components (such as controllers, models, views, helpers, jobs)
+  - commonly used classes in the application that have a common purpose and
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:171]: Sure, can certainly try. Any suggestions?

File: source/manual/conventions-in-rails-applications.html.md [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    interface (such as services or policies)
+- `/lib` is used for:
+  - classes that don't logically fit into an app directory
+  - complex business logic for the application
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:175]: Yup, do you have anything in mind? my go-to is normally Publishing API link expansion code.

File: source/manual/conventions-in-rails-applications.html.md [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+different abstractions (unit, integration and functional)][cp-testing-strategy].
+
+### Common tooling
+
+The preferred framework for testing Rails applications is [rspec-rails][] where
+we aim to adhere to the projects conventions. [rubocop-govuk][] provides
+a linting configuration for RSpec. For mocking and test doubles you should use
+the provided [rspec-mocks][] framework.
+
+For mocking HTTP requests [webmock][] is the common choice with us conventionally
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:238]: Oh yes, good point.

File: source/manual/conventions-in-rails-applications.html.md (hunk 3 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:231]: Ooh good catch - where do they come from?

File: source/manual/conventions-in-rails-applications.html.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/conventions-in-rails-applications.html.md:175]: Sure, I think Content Publisher has a few examples:

- Requirements
- Whitehall Importer
- Versioning
- Bulk Data

File: source/manual/conventions-in-rails-applications.html.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/conventions-in-rails-applications.html.md:171]: I think there are quite a few good examples, although they tend to be "presenters", which perhaps isn't something we want to encourage. Maybe "policies" or even "interactors"?

Bad example: https://github.com/alphagov/specialist-publisher/tree/d6b489d66641de4cd4c557e08ac2c7d70deaceab/app/reporters

File: source/manual/conventions-in-rails-applications.html.md (hunk 3 above)
Comment (ChrisBAshton) [reply source/manual/conventions-in-rails-applications.html.md]: It happens to me a lot too! It's the result of `ALT + SPACE`, which is common because the `ALT` key is often still being held down after pressing the 3 key to make a `#` for the heading.

File: source/manual/conventions-in-rails-applications.html.md [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+[adr-example]: https://github.com/alphagov/content-publisher/blob/master/docs/adr/0002-use-local-datastore.md
+
+## Editing of microcopy by non-developers
+
+When it is intended for non-developers to edit microcopy in an application it is
+conventional to use [Rails Internationalization (I18n)][i18n] to abstract
+the copy outside of the application code. We use this even when an application
+is only in English. These files are stored in `/config/locales` as per Rails
+convention. You should be cautious to only these for accessing copy and not
+for any application configuration.
```
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md:206]: I feel like perhaps this isn't the best example as it implies a technical reason prevents us from using it for configuration when, more strictly, it's a separation of concerns issue.

File: source/manual/conventions-in-rails-applications.html.md (hunk 3 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: Oh thank you! TIL

Comment (issyl0) [reply source/manual/conventions-in-rails-applications.html.md]: I've got this in my `.vimrc` to auto-replace them: https://github.com/issyl0/dotfiles/blob/master/.vimrc#L42-L45

**Comment by kevindew** [comment]
Thanks for the review @benthorner 👍 I've quickly fixed the issues that didn't require me to do much thinking/digging.

I do like your idea about the structure of the file, it's a good idea and does feel logical. I also think it would come at the expense of some clarity for things that end up somewhat broader and cross cutting (ie data storage is more than just a model concern - we don't tend to see Redis models - and also a configuration concern). So does feel like logically something to iterate on after initial pass. 

I'd wonder if you'd not want to be fully rigid with directory structure but more have sections grouped around themes: 

- domain logic (`app` / `lib` directories)
- configuration (`config`)
- testing (`spec` directory)
- dependencies

which could follow the structure but be loss dogmatic than the particular files 🤔

File: source/manual/conventions-in-rails-applications.html.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply source/manual/conventions-in-rails-applications.html.md:206]: Sure. I was motivated [by a specific example](https://github.com/alphagov/collections/blob/master/config/locales/en/transition_landing_page.yml#L16), where everything is a locale, but the data structure has arrays. Conversely, I guess this point is about not using integer or non-printable values in locale files?

File: source/manual/conventions-in-rails-applications.html.md (hunk 3 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: Ooh pro tip 👍

File: source/manual/conventions-in-rails-applications.html.md [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+[request specs][], which have been the [recommended
+approach][rspec-request-moj] to replace [controller specs][] reflecting the
+[direction of Rails 5][controller-rails-5].
+
+[request specs]: https://relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec
+[rspec-request-moj]: https://medium.com/just-tech/rspec-controller-or-request-specs-d93ef563ef11
+[controller specs]: https://relishapp.com/rspec/rspec-rails/docs/controller-specs
+[controller-rails-5]: https://github.com/rails/rails/issues/18950
+
+### Javascript
```
Comment (alex-ju) [inline source/manual/conventions-in-rails-applications.html.md:281]: I share the same view on JavaScript testing 👍

**Comment by benthorner** [comment]
@kevindew ah yes, I see what you mean - I completely skipped over `db/`, for example! Yes, having a top layer of headings would be helpful. Using those same trigger words in headings (like "Gemfile") where possible, too.

> domain logic (app / lib directories)

People seem to have different interpretations of "domain". One interpretation involves "not putting domain code in lib/". Maybe just "app code", assuming tests and config are not code?

File: source/manual/conventions-in-rails-applications.html.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply source/manual/conventions-in-rails-applications.html.md:171]: What's wrong with presenters?

[the remark this answers is not in the document]
Comment (leenagupte) [reply source/manual/conventions-in-rails-applications.html.md:171]: As for examples, smart-answers have an _interesting_ layout: https://github.com/alphagov/smart-answers.

File: source/manual/conventions-in-rails-applications.html.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply source/manual/conventions-in-rails-applications.html.md:206]: I think saying "no arrays" is too restrictive. I18n files are used to define radio button and checkbox content which make sense to be structured as an array.

**Comment by leenagupte** [comment]
This all makes sense to me. Is there a plan to link to this doc from anywhere? i.e. https://docs.publishing.service.gov.uk/manual/setting-up-new-rails-app.html or in the README for https://github.com/alphagov/govuk_app_config?

File: source/manual/conventions-in-rails-applications.html.md [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+[Ruby on Rails]: https://rubyonrails.org/
+[consistent-govuk]: https://docs.google.com/document/d/1jEjPOFJ2s1cjQv9vHAbE-KF68LSletpnUVaG5lXlHy4/edit#heading=h.yod40rauhyhu
+
+## Follow industry conventions
+
+GOV.UK's use of Rails is intended to be consistent with the conventions of
+the framework itself and we intend to embrace developments and deprecations
+from the framework. Where GOV.UK apps differ from Rails conventions (such as
+using the RSpec testing framework) these are intended to be consistent with
+common industry practices.
```
Comment (cbaines) [inline source/manual/conventions-in-rails-applications.html.md:36]: I guess when I read this, maybe you mean something more like "best practices", or actually common practice, which I guess means the same as popular practice, so do it how most other people do. Which was more your intent?

Personally, my view is that using tools in accordance with their affordances is beneficial, but it's encouraged to take different approaches when it's discussed and supported by some reasoning/evidence.

Taking the example of RSpec further, that's not the Rails default, and Basecamp, who I would see as setting industry conventions with regards to Rails (as they're the main industry body related to it) don't use RSpec as far as I'm aware, but again I'm personally all for making this less about what others use, and more about what the people working on GOV.UK think would be better for GOV.UK.

File: source/manual/conventions-in-rails-applications.html.md (hunk 1 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: I've added a quick bit of info to this in 5ef55c07

File: source/manual/conventions-in-rails-applications.html.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: I've added some brief why in 5ef55c07

File: source/manual/conventions-in-rails-applications.html.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: Updated

File: source/manual/conventions-in-rails-applications.html.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: I've updated the doc with a couple of examples.

@leenagupte Rather than presenters as a whole being bad, our GOV.UK use of them is inconsistent which makes them somewhat meaningless as a concept. The main problem I'd suggest we have is that what we call a "presenter" can mean a variety of things, which isn't the same for say controllers, views, models, etc and can often reveal smells. Our use that is closest to the [Model-View-Presenter](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) pattern (and thus, arguably least surprising) is perhaps [GOV.UK Publishing Components](https://github.com/alphagov/govuk_publishing_components/blob/4d4119e0e3424ceefbfb906416ca51d5e3466a03/lib/govuk_publishing_components/presenters/attachment.rb) or [government-frontend](https://github.com/alphagov/government-frontend/tree/e2371b541a7248600099250aed565dfc27233c15/app/presenters) where they act as a model for presentation. However, in other apps we use them for serialisation ([Publishing API](https://github.com/alphagov/publishing-api/blob/eccb8548b286a69adfb83c91f8c74603273a956b/app/presenters/edition_presenter.rb), [Email Alert API](https://github.com/alphagov/email-alert-api/blob/c6fe159dc943deef23d155095996868166624447/app/presenters/content_change_presenter.rb)) and this seems to leave a bit more of a smell as they don't have anything in common with the other type of presenters and are often named after a particular model when they're actually serialising for a particular context.

File: source/manual/conventions-in-rails-applications.html.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: Thanks for the examples, I went with my Publishing API one to balance out the Content Publisher dominance

File: source/manual/conventions-in-rails-applications.html.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: I haven't amended this. I think it sufficiently covered by saying don't use configuration 🤞

File: source/manual/conventions-in-rails-applications.html.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: I've added a note

File: source/manual/conventions-in-rails-applications.html.md (hunk 12 above)
Comment (kevindew) [reply source/manual/conventions-in-rails-applications.html.md]: Thanks for commenting. 👍 

I guess I'm hoping for common and good practices. I'm not sure that outside of Rails' own guides there are clear best practice, so forgoing that I'd suggest things that are commonly done that, collectively, people tend to agree are good (or at least not anti-patterns).

There are some things that may produce a really good result, such as our (internally controversial) use of hexagonal architecture, that in some circles are a best practice but, because it's not a particularly common approach, can be surprising. This can make it difficult for developers familiar with Rails' conventions to learn those applications. 

I think the advantage in being somewhat consistent with Basecamp and wider Rails industry is that we're continuing in the same stream as others and thus are more likely to find our tooling supported, maintained and compatible with future developments.

**Comment by kevindew** [comment]
> This all makes sense to me. Is there a plan to link to this doc from anywhere? i.e. https://docs.publishing.service.gov.uk/manual/setting-up-new-rails-app.html or in the README for https://github.com/alphagov/govuk_app_config?

Thanks @leenagupte, yeah I think it could do with a link on setting up a new rails app. In my testing it does appear autolinked, but not sure how predictable that is (or if anyone reads that far). I'll add a commit for it. I don't think it quite fits into the govuk_app_config README.

@benthorner I applied a bit more structure to the document in a commit. It's not as exciting as your proposal but does group major themes together. Hopefully a modest improvement.

I included the `spec_helper.rb` point. I haven't included anything for validation as I don't think there's a clear convention to follow, what we've done in Content Publisher is good but it seems somewhat bespoke.

**Comment by bilbof** [comment]
@kevindew great work writing this up. The old doc has strong opinions about using the old secrets.yml file, providing secrets via environment variables, and not using the rails credentials feature. I think we should still recommend this.

File: source/manual/conventions-in-rails-applications.html.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply source/manual/conventions-in-rails-applications.html.md]: OK, so it's still ok to use Presenters as long as it's for their intended purpose of reading data to be rendered in a view?

**Comment by kevindew** [comment]
> @kevindew great work writing this up. The old doc has strong opinions about using the old secrets.yml file, providing secrets via environment variables, and not using the rails credentials feature. I think we should still recommend this.

Which old doc do you mean? (apologies if I'm missing something obvious)

There's a chunk on this theme in [the 12 factor section](https://github.com/alphagov/govuk-developer-docs/blob/rails-app-conventions/source/manual/conventions-for-rails-applications.html.md#embrace-12-factor-conventions) but doesn't go explicitly into secrets.

**Comment by bilbof** [comment]
@kevindew sorry I was looking at the bootstrapping doc for rails apps https://docs.publishing.service.gov.uk/manual/setting-up-new-rails-app.html#bootstrapping-a-new-application

**Comment by kevindew** [comment]
Sorry for taking ages to finish this off. Thanks for the suggestions @benthorner I've applied them. 

@bilbof I've put in a brief section on secrets. It wasn't super easy to write as I think our `secrets.yml` approach is mostly because that used to be the norm rather than a strongly opinionated choice. It still seems impractical for us to adopt the `credentials.yml.enc` approach but we may find Rails deprecates the `secrets.yml` approach and we need to consider it more seriously. These seemed somewhat loose reasons for a conventions so I went for a more explicit phrasing.

Let me know if I've dropped any clangers here, I plan to rebase before merging - but thought a distinct commit would be easier to see the changes.

**Comment by bilbof** [comment]
@kevindew thank you! What you've written looks good 👍.
